sharkfuser/include/fusilli/node/matmul_node.h

```diff
@@ -0,0 +1,182 @@
+// shapes. For matrices A [..., M, K] and B [..., K, N], the output is [..., M,
+// N].
+inline std::vector<int64_t>
+getMatmulInferredOutputShape(const std::vector<int64_t> &aDim,
+                             const std::vector<int64_t> &bDim) {
+  size_t aRank = aDim.size();
+  size_t bRank = bDim.size();
+  size_t outRank = std::max(aRank, bRank);
+
+  std::vector<int64_t> cDim(outRank);
+
+  // Handle batch dimensions (broadcast if necessary)
+  size_t batchDims = outRank - 2;
+  for (size_t i = 0; i < batchDims; ++i) {
+    int64_t aDimVal = (i < aRank - 2) ? aDim[i] : 1;
+    int64_t bDimVal = (i < bRank - 2) ? bDim[i] : 1;
+    // Use the maximum of the two dimensions (broadcasting rule)
+    cDim[i] = std::max(aDimVal, bDimVal);
+  }
+
+  // Matrix dimensions: M from A, N from B
+  cDim[outRank - 2] = aDim[aRank - 2]; // M
+  cDim[outRank - 1] = bDim[bRank - 1]; // N
+
+  return cDim;
+}
+
+//===----------------------------------------------------------------------===//
+// Matrix multiplication node.
+//===----------------------------------------------------------------------===//
+
+class MatmulNode : public NodeCRTP<MatmulNode> {
+public:
+  MatmulAttr matmulAttr;
+
+  MatmulNode(MatmulAttr &&attr, const Context &ctx)
+      : NodeCRTP(ctx), matmulAttr(std::move(attr)) {}
+
+  const std::string &getName() const override final {
+    return matmulAttr.getName();
+  }
+  Type getType() const override final { return Type::Matmul; }
+
+  ErrorObject preValidateNode() const override final {
+    FUSILLI_LOG_LABEL_ENDL("INFO: Pre-Validating MatmulNode '"
+                           << matmulAttr.getName() << "'");
+
+    std::shared_ptr<TensorAttr> aT = matmulAttr.getA();
+    std::shared_ptr<TensorAttr> bT = matmulAttr.getB();
+    std::shared_ptr<TensorAttr> cT = matmulAttr.getC();
+
+    // Ensure input and output tensors are set.
+    FUSILLI_RETURN_ERROR_IF(!aT, ErrorCode::AttributeNotSet,
+                            "Matmul input tensor A not set");
+    FUSILLI_RETURN_ERROR_IF(!bT, ErrorCode::AttributeNotSet,
+                            "Matmul input tensor B not set");
+    FUSILLI_RETURN_ERROR_IF(!cT, ErrorCode::AttributeNotSet,
+                            "Matmul output tensor C not set");
+
+    size_t aRank = aT->getDim().size();
+    size_t bRank = bT->getDim().size();
+
+    // Rank checks on input tensors (must be at least rank 2).
+    FUSILLI_RETURN_ERROR_IF(
+        aRank < 2, ErrorCode::InvalidAttribute,
+        "Matmul input tensor A must have a rank of at least 2");
+    FUSILLI_RETURN_ERROR_IF(
+        bRank < 2, ErrorCode::InvalidAttribute,
+        "Matmul input tensor B must have a rank of at least 2");
+
+    // Check that inner dimensions match (K dimension).
+    const std::vector<int64_t> &aDim = aT->getDim();
+    const std::vector<int64_t> &bDim = bT->getDim();
+    int64_t aK = aDim[aRank - 1]; // Last dimension of A
+    int64_t bK = bDim[bRank - 2]; // Second-to-last dimension of B
```

Comment:
Does this assume a specific stride order, or is it sufficient to say the logical dims are always "row-major" / "contiguous" for this to be correct?